aws-lc-rs/src/cipher/key.rs

```diff
@@ -0,0 +1,200 @@
+            }
+            let dec_key = dec_key.assume_init();
+
+            let mut kb = MaybeUninit::<[u8; 16]>::uninit();
+            ptr::copy_nonoverlapping(key_bytes.as_ptr(), kb.as_mut_ptr().cast(), 16);
+            Ok(SymmetricCipherKey::Aes128 {
+                raw_key: Aes128Key(kb.assume_init()),
+                enc_key,
+                dec_key,
+            })
+        }
+    }
+
+    pub(crate) fn aes256(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != 32 {
+            return Err(Unspecified);
+        }
+        unsafe {
+            let mut enc_key = MaybeUninit::<AES_KEY>::uninit();
+            #[allow(clippy::cast_possible_truncation)]
+            if 0 != AES_set_encrypt_key(
+                key_bytes.as_ptr(),
+                (key_bytes.len() * 8) as c_uint,
+                enc_key.as_mut_ptr(),
+            ) {
+                return Err(Unspecified);
+            }
+            let enc_key = enc_key.assume_init();
+
+            let mut dec_key = MaybeUninit::<AES_KEY>::uninit();
+            #[allow(clippy::cast_possible_truncation)]
+            if 0 != AES_set_decrypt_key(
+                key_bytes.as_ptr(),
+                (key_bytes.len() * 8) as c_uint,
+                dec_key.as_mut_ptr(),
+            ) {
+                return Err(Unspecified);
+            }
+            let dec_key = dec_key.assume_init();
+
+            let mut kb = MaybeUninit::<[u8; 32]>::uninit();
+            copy_nonoverlapping(key_bytes.as_ptr(), kb.as_mut_ptr().cast(), 32);
+            Ok(SymmetricCipherKey::Aes256 {
+                raw_key: Aes256Key(kb.assume_init()),
+                enc_key,
+                dec_key,
+            })
+        }
+    }
+
+    pub(crate) fn chacha20(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != 32 {
+            return Err(Unspecified);
+        }
+        let mut kb = MaybeUninit::<[u8; 32]>::uninit();
+        unsafe {
+            copy_nonoverlapping(key_bytes.as_ptr(), kb.as_mut_ptr().cast(), 32);
+            Ok(SymmetricCipherKey::ChaCha20 {
+                raw_key: ChaCha20Key(kb.assume_init()),
+            })
+        }
+    }
+
+    #[inline]
+    pub(crate) fn key_bytes(&self) -> &[u8] {
+        match self {
+            SymmetricCipherKey::Aes128 { raw_key, .. } => &raw_key.0,
+            SymmetricCipherKey::Aes256 { raw_key, .. } => &raw_key.0,
+            SymmetricCipherKey::ChaCha20 { raw_key, .. } => &raw_key.0,
+        }
+    }
+
+    #[allow(dead_code)]
+    #[inline]
+    pub(crate) fn encrypt_block(&self, block: Block) -> Block {
+        match self {
+            SymmetricCipherKey::Aes128 { enc_key, .. }
+            | SymmetricCipherKey::Aes256 { enc_key, .. } => encrypt_block_aes(enc_key, block),
+            SymmetricCipherKey::ChaCha20 { .. } => panic!("Unsupported algorithm!"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cipher::block::{Block, BLOCK_LEN};
```

Comment:
Is `BLOCK_LEN` redundant with `AES_BLOCK_LEN`, or do they imply different things?

Reply:
Both the type `Block` and the value `BLOCK_LEN` are only internally visible (`pub(crate)`). The type is primarily used for [`aead::HeaderProtectionKey::new_mask`](https://docs.rs/aws-lc-rs/latest/aws_lc_rs/aead/quic/struct.HeaderProtectionKey.html#method.new_mask) which uses a "sample_len" of 16 (even for ChaCha20), which also happens to align with `BLOCK_LEN`.